### PR TITLE
Update coldsnap to use aws sdk for rust

### PR DESCRIPTION
**Pull request by rpkelly**
Removed rusoto, updated dependencies for aws-sdk-rust, and updated
functions to call the AWS sdk.

Removed the client mod, it was no longer necessary with how clients are
built in the AWS sdk.

*Description of changes:*
Removed rusoto and replaced it with the AWS SDK Rust.

Ran the following tests:
* Unit tests were updated to use the Rust SDK
* Generated 2 GB random file, uploaded, downloaded and confirmed that hash remained the same
* Confirmed that profile and region settings work as intended

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -10,23 +10,23 @@ repository = "https://github.com/awslabs/coldsnap"
...
 
 [dependencies]
 argh = "0.1.7"
-tokio = { version = "1.8", features = ["fs", "io-util", "time"] }
+aws-config = "0.11.0"
+aws-sdk-ebs = "0.11.0"
+aws-sdk-ec2 = "0.11.0"
+aws-smithy-http = "0.41.0"
+aws-types = "0.11.0"
+tokio = { version = "~1.14", features = ["full"] }
```
Comment (webern) [inline Cargo.toml]: Cargo.toml files that are used in libraries should not lock in a major.minor version because they can make it impossible for a client to build (if some other library requires some other minor version).

I also think that our libraries should not require a minimum tokio version unless there is some specific reason for it. If we wanted to lock to a different version in the Bottlerocket tools project with, e.g. `~1.8`, then the I think it fail because that would be lower than this one.

```suggestion
tokio = { version = "1", features = ["full"] }
```

File: src/bin/coldsnap/main.rs [hunk 2]
```diff
@@ -35,9 +32,59 @@ async fn main() {
...
     let args: Args = argh::from_env();
+    let config: aws_config::ConfigLoader = match (args.region, &args.profile) {
+        (Some(region), Some(_profile)) => {
+            //region option passed in
+            aws_config::from_env().region(Region::new(region))
+        }
+        (Some(region), None) => {
+            //region option passed in
+            aws_config::from_env().region(Region::new(region))
+        }
```
Comment (webern) [inline src/bin/coldsnap/main.rs]: I think this works (Also nit, strive for comment format consistency: a space after the `//` and it looks like other comments in the lib start with upper case):

```suggestion
        (Some(region), _) => {
            // Region option passed in
            aws_config::from_env().region(Region::new(region))
        }
```

File: src/download.rs [hunk 3]
```diff
@@ -277,11 +272,18 @@ impl SnapshotDownloader {
...
-        let block_data = response.block_data.context(error::FindBlockPropertySnafu {
-            snapshot_id,
-            block_index,
-            property: "data",
-        })?;
+        let block_data_stream = response
+            .block_data
+            .collect()
+            .await
+            //.map(|data| data.into_bytes())
```
Comment (webern) [inline src/download.rs]:
```suggestion
```

File: src/wait.rs [hunk 4]
```diff
@@ -148,16 +147,16 @@ impl SnapshotWaiter {
...
     use snafu::Snafu;
 
     #[derive(Debug, Snafu)]
     #[snafu(visibility(pub(super)))]
     pub(super) enum Error {
         #[snafu(display("Failed to describe snapshots: {}", source))]
         DescribeSnapshots {
-            source: RusotoError<DescribeSnapshotsError>,
+            #[snafu(source(from(aws_sdk_ec2::types::SdkError<DescribeSnapshotsError>, Box::new)))]
+            source: Box<aws_sdk_ec2::types::SdkError<DescribeSnapshotsError>>,
```
Comment (webern) [inline src/wait.rs:161]: Is the box necessary? (Was it a clippy issue with the size of the error? If so a comment would be nice to explain why we boxed this one error type)

```suggestion
            source: aws_sdk_ec2::types::SdkError<DescribeSnapshotsError>,
```

File: src/upload.rs [hunk 5]
```diff
@@ -389,18 +393,19 @@ impl SnapshotUploader {
...
     data_length: usize,
-    block_index: i64,
-    block_size: i64,
+    block_index: i32,
+    block_size: i32,
     snapshot_id: String,
-    changed_blocks_count: Arc<AtomicI64>,
-    block_digests: Arc<Mutex<BTreeMap<i64, Vec<u8>>>>,
-    block_errors: Arc<Mutex<BTreeMap<i64, Error>>>,
+    changed_blocks_count: Arc<AtomicI32>,
```
Comment (webern) [inline src/upload.rs:395]: I wonder how large these numbers get. It looks like the rusoto codegen and the new aws sdk codegen chose different sizes for integers. i64 seems like a safer choice. Nothing we can do about it in this PR other than considering the maximum values we might expect for these to anticipate whether there is a problem or not.

The AWS Rust SDK team could probably change the codegen use i64 instead if its a problem.

```
    changed_blocks_count: Arc<AtomicI32>,
```

File: src/upload.rs [hunk 6]
```diff
@@ -133,7 +140,12 @@ impl SnapshotUploader {
             // The file length may not be an exact multiple of the block size,
             // so we need to keep track of how many bytes are left for the call
             // to `read_exact` later.
-            let data_length = cmp::min(block_size, remaining_data);
+            let data_length = cmp::min(i64::from(block_size), remaining_data);
+            let data_length = i32::try_from(data_length).context(error::ConvertNumberSnafu {
```
Comment (webern) [inline src/upload.rs]: Good that you have handled a possible narrowing error, see my i32 concern elsewhere.

File: src/bin/coldsnap/main.rs [hunk 7]
```diff
@@ -35,9 +32,59 @@ async fn main() {
...
+                aws_config::profile::ProfileFileCredentialsProvider::builder()
+                    .profile_name(profile)
+                    .build(),
+            )
+        }
+        None => {
+            // keep config equal to itself
+            config
+        }
+    };
```
Comment (webern) [inline src/bin/coldsnap/main.rs]: Slightly golfed. Are the type annotations necessary?

```suggestion
    let config = if let Some(profile) = match &args.profile {
            // Add profile credential provider
            config.credentials_provider(
                aws_config::profile::ProfileFileCredentialsProvider::builder()
                    .profile_name(profile)
                    .build(),
            )
    } else {
        // Keep config unchanged
        config
    };
```

**Review by webern — COMMENTED** [review]
Do we need to try it in other partitions and/or opt-in regions?

File: Cargo.toml (hunk 1 above)
Comment (rpkelly) [reply Cargo.toml]: My understanding is that we intentionally lock to the LTS version of tokio. Previously that had been 1.8.x, now it's 1.14.x. Is that not true?

File: src/bin/coldsnap/main.rs (hunk 2 above)
Comment (rpkelly) [reply src/bin/coldsnap/main.rs]: Thanks!

File: src/wait.rs (hunk 4 above)
Comment (rpkelly) [reply src/wait.rs:161]: Yeah, it was a clippy size thing, I'll add a comment

File: src/upload.rs (hunk 5 above)
Comment (rpkelly) [reply src/upload.rs:395]: Block size is locked to 512 kibibytes (for now).

(2,147,483,647 * 512)/(1024^2) =  1,048,575.99951 gibibytes can be changed before we run out of space. 

This seems safe, at least for the foreseeable future.

File: src/upload.rs (hunk 6 above)
Comment (rpkelly) [reply src/upload.rs]: Funnily enough, since block size is locked to 512 kb, this is actually a relatively safe conversation.

**Review by webern — APPROVED** [review]
We might to double-check that it works in other partitions (i.e. double-check that the AWS Rust SDK works the way we expect to). Other than that looks good!

File: Cargo.toml [hunk 8]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "coldsnap"
-version = "0.3.3"
```
Comment (jpculp) [inline Cargo.toml]: We should probably leave the version bump for a separate PR once we have a changelog and we're ready to release a new crate.

**Review by jpculp — CHANGES_REQUESTED** [review]

**Review by zmrow — APPROVED** [review]
:lollipop:

File: Cargo.toml [hunk 9]
```diff
@@ -10,23 +10,23 @@ repository = "https://github.com/awslabs/coldsnap"
...
 
 [dependencies]
 argh = "0.1.7"
-tokio = { version = "1.8", features = ["fs", "io-util", "time"] }
+aws-config = "0.11.0"
+aws-sdk-ebs = "0.11.0"
+aws-sdk-ec2 = "0.11.0"
+aws-smithy-http = "0.41.0"
+aws-types = "0.11.0"
+tokio = { version = "1", features = ["full"] }
```
Comment (bcressey) [inline Cargo.toml]: Do we need all the features of `tokio`, or can we use our usual set?
```suggestion
tokio = { version = "~1.14", default-features = false, features = ["macros", "rt-multi-thread"] }  # LTS
```

File: deny.toml [hunk 10]
```diff
@@ -42,8 +42,10 @@ multiple-versions = "deny"
 wildcards = "deny"
 
 skip-tree = [
-    # rusoto_signature uses an older version of sha2
-    { name = "rusoto_signature" },
+    # h2 pulls in an older version of tokio util
+    { name = "h2",  version = "0.3.13" },
```
Comment (bcressey) [inline deny.toml]: This happens because the AWS SDK depends on both tokio and h2, and h2 depends on an older version of tokio?

File: src/upload.rs [hunk 11]
```diff
@@ -98,10 +100,15 @@ impl SnapshotUploader {
 
         // Start the snapshot, which gives us the ID and block size we need.
         let (snapshot_id, block_size) = self.start_snapshot(volume_size, description).await?;
-        let file_blocks = (file_size + block_size - 1) / block_size;
+        let file_blocks = (file_size + i64::from(block_size - 1)) / i64::from(block_size);
+        let file_blocks = i32::try_from(file_blocks).context(error::ConvertNumberSnafu {
```
Comment (bcressey) [inline src/upload.rs]: It's better to use `with_context` for lazy generation of the struct.

File: src/upload.rs [hunk 12]
```diff
@@ -133,7 +140,12 @@ impl SnapshotUploader {
             // The file length may not be an exact multiple of the block size,
             // so we need to keep track of how many bytes are left for the call
             // to `read_exact` later.
-            let data_length = cmp::min(block_size, remaining_data);
+            let data_length = cmp::min(i64::from(block_size), remaining_data);
+            let data_length = i32::try_from(data_length).context(error::ConvertNumberSnafu {
+                what: "data length",
+                number: data_length.to_string(),
+                target: "i32",
+            })?;
```
Comment (bcressey) [inline src/upload.rs]: Why do we need to convert an i64 to an i32 and then to a usize, instead of converting it directly to a usize?

File: src/wait.rs [hunk 13]
```diff
@@ -118,7 +117,7 @@ impl SnapshotWaiter {
                 for snapshot in snapshots {
                     if let Some(ref found_id) = snapshot.snapshot_id {
                         if let Some(ref found_state) = snapshot.state {
-                            if snapshot_id.as_ref() == found_id && &state == found_state {
+                            if snapshot_id.as_ref() == found_id && state == found_state.as_str() {
```
Comment (bcressey) [inline src/wait.rs]: If you change this to `if let Some(found_state)` then you can avoid the dereference change below - seems like we don't actually want a ref now.

File: src/bin/coldsnap/main.rs [hunk 14]
```diff
@@ -35,9 +32,52 @@ async fn main() {
 
 async fn run() -> Result<()> {
     let args: Args = argh::from_env();
+    let config: aws_config::ConfigLoader = match (args.region, &args.profile) {
```
Comment (bcressey) [inline src/bin/coldsnap/main.rs]: Can we move this all into a helper function that returns a `ConfigLoader`?

File: src/bin/coldsnap/main.rs [hunk 15]
```diff
@@ -35,9 +32,52 @@ async fn main() {
...
+        (Some(region), _) => {
+            // Region option passed in
+            aws_config::from_env().region(Region::new(region))
```
Comment (bcressey) [inline src/bin/coldsnap/main.rs]: I use an EC2 instance for development that has a fairly limited instance role assigned, and an `~/.aws/config` file with a more powerful profile for use as needed.

Just want to confirm that this scenario still works.

File: Cargo.toml (hunk 9 above)
Comment (rpkelly) [reply Cargo.toml]: aws sdk rust specifies "full" https://github.com/awslabs/aws-sdk-rust. I'm unsure of the consequences of not using full, so I followed the docs

File: deny.toml (hunk 10 above)
Comment (rpkelly) [reply deny.toml]:
```
tokio-util v0.6.10
      └── aws-smithy-http v0.41.0
          ├── aws-config v0.11.0
          │   └── coldsnap v0.4.0
          ├── aws-endpoint v0.11.0
          │   ├── aws-sdk-ebs v0.11.0
          │   │   └── coldsnap v0.4.0 (*)
          │   ├── aws-sdk-ec2 v0.11.0
          │   │   └── coldsnap v0.4.0 (*)
          │   ├── aws-sdk-sso v0.11.0
          │   │   └── aws-config v0.11.0 (*)
          │   └── aws-sdk-sts v0.11.0
          │       └── aws-config v0.11.0 (*)
          ├── aws-http v0.11.0
          │   ├── aws-config v0.11.0 (*)
          │   ├── aws-sdk-ebs v0.11.0 (*)
          │   ├── aws-sdk-ec2 v0.11.0 (*)
          │   ├── aws-sdk-sso v0.11.0 (*)
          │   └── aws-sdk-sts v0.11.0 (*)
          ├── aws-sdk-ebs v0.11.0 (*)
          ├── aws-sdk-ec2 v0.11.0 (*)
          ├── aws-sdk-sso v0.11.0 (*)
          ├── aws-sdk-sts v0.11.0 (*)
          ├── aws-sig-auth v0.11.0
          │   ├── aws-sdk-ebs v0.11.0 (*)
          │   ├── aws-sdk-ec2 v0.11.0 (*)
          │   ├── aws-sdk-sso v0.11.0 (*)
          │   └── aws-sdk-sts v0.11.0 (*)
          ├── aws-sigv4 v0.11.0
          │   └── aws-sig-auth v0.11.0 (*)
          ├── aws-smithy-client v0.41.0
          │   ├── aws-config v0.11.0 (*)
          │   ├── aws-sdk-ebs v0.11.0 (*)
          │   ├── aws-sdk-ec2 v0.11.0 (*)
          │   ├── aws-sdk-sso v0.11.0 (*)
          │   ├── aws-sdk-sts v0.11.0 (*)
          │   └── aws-types v0.11.0
          │       ├── aws-config v0.11.0 (*)
          │       ├── aws-endpoint v0.11.0 (*)
          │       ├── aws-http v0.11.0 (*)
          │       ├── aws-sdk-ebs v0.11.0 (*)
          │       ├── aws-sdk-ec2 v0.11.0 (*)
          │       ├── aws-sdk-sso v0.11.0 (*)
          │       ├── aws-sdk-sts v0.11.0 (*)
          │       ├── aws-sig-auth v0.11.0 (*)
          │       └── coldsnap v0.4.0 (*)
          ├── aws-smithy-http-tower v0.41.0
          │   ├── aws-config v0.11.0 (*)
          │   ├── aws-sdk-ebs v0.11.0 (*)
          │   ├── aws-sdk-ec2 v0.11.0 (*)
          │   ├── aws-sdk-sso v0.11.0 (*)
          │   ├── aws-sdk-sts v0.11.0 (*)
          │   └── aws-smithy-client v0.41.0 (*)
          ├── aws-types v0.11.0 (*)
          └── coldsnap v0.4.0 (*)
    = tokio-util v0.7.2
      └── h2 v0.3.13
          └── hyper v0.14.18
              ├── aws-config v0.11.0
              │   └── coldsnap v0.4.0
              ├── aws-smithy-client v0.41.0
              │   ├── aws-config v0.11.0 (*)
              │   ├── aws-sdk-ebs v0.11.0
              │   │   └── coldsnap v0.4.0 (*)
              │   ├── aws-sdk-ec2 v0.11.0
              │   │   └── coldsnap v0.4.0 (*)
              │   ├── aws-sdk-sso v0.11.0
              │   │   └── aws-config v0.11.0 (*)
              │   ├── aws-sdk-sts v0.11.0
              │   │   └── aws-config v0.11.0 (*)
              │   └── aws-types v0.11.0
              │       ├── aws-config v0.11.0 (*)
              │       ├── aws-endpoint v0.11.0
              │       │   ├── aws-sdk-ebs v0.11.0 (*)
              │       │   ├── aws-sdk-ec2 v0.11.0 (*)
              │       │   ├── aws-sdk-sso v0.11.0 (*)
              │       │   └── aws-sdk-sts v0.11.0 (*)
              │       ├── aws-http v0.11.0
              │       │   ├── aws-config v0.11.0 (*)
              │       │   ├── aws-sdk-ebs v0.11.0 (*)
              │       │   ├── aws-sdk-ec2 v0.11.0 (*)
              │       │   ├── aws-sdk-sso v0.11.0 (*)
              │       │   └── aws-sdk-sts v0.11.0 (*)
              │       ├── aws-sdk-ebs v0.11.0 (*)
              │       ├── aws-sdk-ec2 v0.11.0 (*)
              │       ├── aws-sdk-sso v0.11.0 (*)
              │       ├── aws-sdk-sts v0.11.0 (*)
              │       ├── aws-sig-auth v0.11.0
              │       │   ├── aws-sdk-ebs v0.11.0 (*)
              │       │   ├── aws-sdk-ec2 v0.11.0 (*)
              │       │   ├── aws-sdk-sso v0.11.0 (*)
              │       │   └── aws-sdk-sts v0.11.0 (*)
              │       └── coldsnap v0.4.0 (*)
              ├── aws-smithy-http v0.41.0
              │   ├── aws-config v0.11.0 (*)
              │   ├── aws-endpoint v0.11.0 (*)
              │   ├── aws-http v0.11.0 (*)
              │   ├── aws-sdk-ebs v0.11.0 (*)
              │   ├── aws-sdk-ec2 v0.11.0 (*)
              │   ├── aws-sdk-sso v0.11.0 (*)
              │   ├── aws-sdk-sts v0.11.0 (*)
              │   ├── aws-sig-auth v0.11.0 (*)
              │   ├── aws-sigv4 v0.11.0
              │   │   └── aws-sig-auth v0.11.0 (*)
              │   ├── aws-smithy-client v0.41.0 (*)
              │   ├── aws-smithy-http-tower v0.41.0
              │   │   ├── aws-config v0.11.0 (*)
              │   │   ├── aws-sdk-ebs v0.11.0 (*)
              │   │   ├── aws-sdk-ec2 v0.11.0 (*)
              │   │   ├── aws-sdk-sso v0.11.0 (*)
              │   │   ├── aws-sdk-sts v0.11.0 (*)
              │   │   └── aws-smithy-client v0.41.0 (*)
              │   ├── aws-types v0.11.0 (*)
              │   └── coldsnap v0.4.0 (*)
              ├── hyper-rustls v0.22.1
              │   └── aws-smithy-client v0.41.0 (*)
              └── hyper-tls v0.5.0
                  └── aws-smithy-client v0.41.0 (*)
                  ```

File: src/bin/coldsnap/main.rs (hunk 15 above)
Comment (rpkelly) [reply src/bin/coldsnap/main.rs]: I called this without having an aws config on an EC2 instance with an instance profile, and also tested the profiles after I created them on the same EC2 instance.

File: src/upload.rs (hunk 12 above)
Comment (rpkelly) [reply src/upload.rs]: i32 is probably unnecessary since we're guaranteed the size will be inside a 32 bit int. I'll remove that conversion.

**Review by jpculp — APPROVED** [review]

File: Cargo.toml (hunk 9 above)
Comment (webern) [reply Cargo.toml]: > Do we need all the features of tokio

In your suggestion you regressed us back to locking the version! I recommend we stick with `1` unless there is a reason to force a version minimum on downstream lib users.

File: deny.toml (hunk 10 above)
Comment (bcressey) [reply deny.toml]: Looks like this was fixed in the recent 0.42.0 release: https://github.com/awslabs/aws-sdk-rust/commit/77e1c51891a654b49a9f49b0a25e55f9443bdecc.

File: src/bin/coldsnap/main.rs [hunk 16]
```diff
@@ -124,6 +120,55 @@ fn build_progress_bar(no_progress: bool, verb: &str) -> Option<ProgressBar> {
     Some(progress_bar)
 }
 
+async fn build_client_config(
```
Comment (bcressey) [inline src/bin/coldsnap/main.rs:124]: nit: could add a docstring

**Review by bcressey — APPROVED** [review]

File: Cargo.toml (hunk 9 above)
Comment (bcressey) [reply Cargo.toml]: I agree with @webern that we don't want to pin the tokio version.

That example in the aws-sdk-rust readme doesn't really persuade me; they might've just picked "full" because it was convenient - easy for developers to add, unlikely to change over time.

It looks like they're being more specific about tokio features in their own crates. I'd like coldsnap to be similarly specific about the subset of tokio that it needs.

Comment (bcressey) [reply Cargo.toml]: Putting that all together, I think we want this:
```
tokio = { version = "1", features = ["fs", "io-util", "time"] }
```

Like it was before, but using `1` instead of `1.8`.

Comment (webern) [reply Cargo.toml]: Side note: if the SDK chose full features, then that's what we're getting anyway. But I agree with the best practice of specifying our minimum feature set. Pretty sure we will need `rt-multi-threaded` and `macros` as well. The silly thing is that we wouldn't know whether we needed them or not because, if the SDK includes `full` then those features are there.

Anyway if this message is confusing, ignore it.

**Review by jpculp — APPROVED** [review]

**Review by bcressey — APPROVED** [review]

**Review by webern — APPROVED** [review]
